icu4c/source/common/uresbund.cpp

```diff
@@ -1910,10 +1910,74 @@ static Resource getTableItemByKeyPath(const ResourceData *pResData, Resource tab
+static void createPath(CharString& path,
+                       const char* resPath,
+                       int32_t     resPathLen,
+                       const char* origResPath,
+                       int32_t     origResPathLen,
+                       const char* inKey,
+                       UErrorCode* status) {
+    // This is a utility function used by ures_getByKeyWithFallback() below.  This function builds a path from
+    // resPath and inKey, returning the result in `path`.  Originally, this function just cleared `path` and
+    // appended resPath and inKey to it, but that caused problems for horizontal inheritance.
+    //
+    // In normal cases, resPath is the same as origResPath, but if ures_getByKeyWithFallback() has followed an
+    // alias, resPath may be different from origResPath.  Not only may the existing path elements be different,
+    // but resPath may also have MORE path elements than origResPath did.  If it does, those additional path
+    // elements SUPERSEDE the corresponding elements of inKey.  So this code counts the number of elements in
+    // resPath and origResPath and, for each path element in resPath that doesn't have a counterpart in origResPath,
+    // deletes a path element from the beginning of inKey.  The remainder of inKey is then appended to
+    // resPath to form the result.  (We're not using uprv_strchr() here because resPath and origResPath may
+    // not be zero-terminated.)
+    path.clear();
+    const char* key = inKey;
+    if (resPathLen > 0) {
+        path.append(resPath, resPathLen, *status);
+        if (U_SUCCESS(*status)) {
+            const char* resPathEnd = resPath + resPathLen;
+            const char* origResPathEnd = origResPath + origResPathLen;
+            const char* resPathPtr = resPath;
+            const char* origResPathPtr = origResPath;
+            
+            while (origResPathPtr < origResPathEnd && resPathPtr < resPathEnd) {
+                while (origResPathPtr < origResPathEnd && *origResPathPtr != RES_PATH_SEPARATOR) {
+                    ++origResPathPtr;
+                }
+                if (origResPathPtr < origResPathEnd && *origResPathPtr == RES_PATH_SEPARATOR) {
+                    ++origResPathPtr;
+                }
+                while (resPathPtr < resPathEnd && *resPathPtr != RES_PATH_SEPARATOR) {
+                    ++resPathPtr;
+                }
+                if (resPathPtr < resPathEnd && *resPathPtr == RES_PATH_SEPARATOR) {
+                    ++resPathPtr;
+                }
+            }
+            while (resPathPtr < resPathEnd && *key != '\0') {
+                while (resPathPtr < resPathEnd && *resPathPtr != RES_PATH_SEPARATOR) {
+                    ++resPathPtr;
+                }
+                if (resPathPtr < resPathEnd && *resPathPtr == RES_PATH_SEPARATOR) {
+                    ++resPathPtr;
+                }
+                while (*key != '\0' && *key != RES_PATH_SEPARATOR) {
```

Comment:
FYI only:
In Java, I had changed the storage of a keyPath into an array of String, one per segment. Simplified a lot of processing. A rough equivalent in C++ would be storing NUL-terminated segments consecutively in one buffer, maybe with an extra terminating NUL or else with the number of segments at the start or elsewhere. Then we could march through here just with uprv_strchr(keyPath, 0), and elsewhere we wouldn't need to copy the path so that we can NUL-terminate segments for strcmp()... Food for thought for 71.

Reply:
The arrays of String in Java definitely are easier to understand, and the actual logic using them easier to follow (and significantly different from the logic in the C++ code).  I kind of like what you're suggesting here, and I wonder if we could go even further to make the actual algorithms match what we're doing on the Java side, but I was _way_ too scared to attempt anything like that for ICU 70.